scipy/spatial/src/distance_pybind.cpp

```diff
@@ -0,0 +1,593 @@
+        return py::array(dtype, out_shape);
+    }
+
+    if (!py::isinstance<py::array>(obj)) {
+        throw py::type_error("out argument must be an ndarray");
+    }
+
+    py::array out = py::cast<py::array>(obj);
+    const auto ndim = out.ndim();
+    const auto shape = out.shape();
+
+    if (ndim != out_shape.size() ||
+        !std::equal(shape, shape + ndim, out_shape.begin())) {
+        throw std::invalid_argument("Output array has incorrect shape.");
+    }
+    if (out.dtype().not_equal(dtype)) {
+        const py::handle& handle = dtype;
+        throw std::invalid_argument("wrong out dtype, expected " +
+                                    std::string(py::str(handle)));
+    }
+    auto pao = reinterpret_cast<PyArrayObject*>(out.ptr());
+    if (!PyArray_ISBEHAVED(pao)) {
+        throw std::invalid_argument(
+            "out array must be aligned, writable and native byte order");
+    }
+    return out;
+}
+
+py::array prepare_single_weight(const py::object& obj, int64_t len) {
+    py::array weight = npy_asarray(obj);
+    if (weight.ndim() != 1) {
+        throw std::invalid_argument("Weights must be a vector (ndim = 1)");
+    } else if (weight.shape(0) != len) {
+        std::stringstream msg;
+        msg << "Weights must have same size as input vector. ";
+        msg << weight.shape(0) << " vs. " << len << ".";
+        throw std::invalid_argument(msg.str());
+    }
+    return weight;
+}
+
+py::dtype common_type(py::dtype type) {
+    return std::move(type);
+}
+
+template <typename... Args>
+py::dtype common_type(const py::dtype& type1, const py::dtype& type2,
+                      const Args&... tail) {
+    return common_type(npy_promote_types(type1, type2), tail...);
+}
+
+int dtype_num(const py::dtype& dtype) {
+    return reinterpret_cast<const PyArray_Descr*>(
+        dtype.ptr())->type_num;
+}
+
+py::dtype promote_type_real(const py::dtype& dtype) {
+    switch (dtype.kind()) {
+    case 'b':
+    case 'i':
+    case 'u': {
+        // Promote integral and boolean types to double
+        return py::dtype::template of<double>();
+    }
+    case 'f': {
+        if (dtype_num(dtype) == NPY_LONGDOUBLE) {
+            return dtype;
+        } else {
+            // TODO: Allow float32 output
+            return py::dtype::template of<double>();
+        }
+    }
+
+    default: {
+        return dtype;
+    }
+    }
+}
+
+#define DISPATCH_DTYPE(dtype, func)                                     \
+    do {                                                                \
+        const py::dtype& type_obj = dtype;                              \
+        switch (dtype_num(type_obj)) {                                  \
+        case NPY_HALF:                                                  \
+        case NPY_FLOAT: /* TODO: Enable scalar_t=float dispatch */      \
+        case NPY_DOUBLE: {                                              \
+            using scalar_t = double;                                    \
```

Comment:
I am confused by this. The block doesn't access scalar_t, and func() has it's own scope. Could you comment how the using line has effect?

Reply:
> I am confused by this. The block doesn't access scalar_t, and func() has it's own scope.

`func` here isn't an object, it's a block of code that happens to be callable like a function because it defines a lambda. So, it's like copying the block of code twice; once with `scalar_t` defined as `double` and once as `long double`.

Reply:
Thanks for the explanation.

The pattern amounts to a 'reversed' type dependency.
Genearting func() twice is exactly the intended use of templates.
```
template <T> struct MyDestination {
   void func(..., ..., ...) {
        do thing with T
   }
}
template <DEST>
void dispatch_dtype(dtype, ..., ..., ...) {
...
case LONGDOUBLE:
DEST<long double>::func(..., ..., ...);
...
}
```

Or something like this?
```
void dispatch_dtype(dtype, longdouble_do, double_do) {

}

dispatch_dtype(dtype, scalar_do<long double>, scalar_do<double>)
```

Reply:
I mean yes, I am of course using templates. And `scalar_t` is the template parameter. The exact code I need is:
```cpp
        switch (dtype_num(dtype)) {                                  
        case NPY_HALF:                                                  
        case NPY_FLOAT:    
        case NPY_DOUBLE: {                                              
            pdist_unweighted<long double>(out, x, f);
            break;                                                      
        }                                                               
        case NPY_LONGDOUBLE: {                                          
            pdist_unweighted<long double>(out, x, f);                   
            break;                                                      
        }                                                               
        default: {                                                      
            const py::handle& handle = dtype;                        
            throw std::invalid_argument(                                
                "Unsupported dtype " + std::string(py::str(handle)));   
        }                                                               
        }    
```
Except I actually need it at least 4 times (pdist + cdist) * (weighted + unweighted). My options are:
- Write the switch out manually 4 times
- Use this little macro trick
- Some complicated template meta-programming

Reply:
I see. Macro is the simplest.

What about rename func to `lambda_expression`? At least the term 'expression' implies it is not just a plain function object but rather there is verbatim macro expansion involved?

Reply:
Good point, I've renamed to `expression` and added a comment.